android/app/src/main/java/org/electroncash/electroncash3/ColdLoad.kt

```diff
@@ -43,22 +51,79 @@ class ColdLoadDialog : AlertDialogFragment() {
 
     private fun updateUI() {
         val currenttext = etTransaction.text
+        val tx: PyObject
+
         //checks if text is blank. further validations can be added here
-        dialog.getButton(AlertDialog.BUTTON_POSITIVE).isEnabled = currenttext.isNotBlank()
+        if (currenttext.isNotBlank()) {
+            dialog.getButton(AlertDialog.BUTTON_POSITIVE).isEnabled = true
+            tx = libTransaction.callAttr("Transaction", etTransaction.text.toString())
+
+            updateStatusText(tx)
+
+            // Check hex transaction signing status
+            if (canSign(tx)) {
+                dialog.getButton(AlertDialog.BUTTON_POSITIVE).setText(R.string.sign)
+                dialog.getButton(AlertDialog.BUTTON_POSITIVE).isEnabled = true
+            } else if (canBroadcast(tx)) {
+                dialog.getButton(AlertDialog.BUTTON_POSITIVE).setText(R.string.send)
+                dialog.getButton(AlertDialog.BUTTON_POSITIVE).isEnabled = true
+            } else {
+                dialog.getButton(AlertDialog.BUTTON_POSITIVE).isEnabled = false
+            }
+        }
     }
 
     // Receives the result of a QR scan.
     override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
         val result = IntentIntegrator.parseActivityResult(requestCode, resultCode, data)
+
+        // Try to decode the QR content as Base43; if that fails, treat it as is
+        val txHex: String = try {
+            baseDecode(result.contents, 43)
+        } catch (e: PyException) {
+            result.contents
+        }
+
         if (result != null && result.contents != null) {
-            etTransaction.setText(result.contents)
+            etTransaction.setText(txHex)
         } else {
             super.onActivityResult(requestCode, resultCode, data)
         }
     }
 
     fun onOK() {
         val tx = libTransaction.callAttr("Transaction", etTransaction.text.toString())
+
+        // If transaction can be broadcasted, broadcast it.
+        // Otherwise, prompt for signing. If the transaction hex is invalid,
+        // the OK button will be disabled, regardless.
+        try {
+            if (canBroadcast(tx)) {
+                broadcastSignedTransaction(tx)
+            } else {
+                signLoadedTransaction()
+            }
+        } catch (e: ToastException) {
+            e.show()
+            return
+        }
+    }
+
+    /**
+     * Sign a loaded transaction.
+     */
+    private fun signLoadedTransaction() {
+        val dialog = SignPasswordDialog()
```

Comment:
I think we must launch the Send dialog to display the transaction details, otherwise the user doesn't know what they're signing. All the fields in the Send dialog except the description would be read-only, including the fee slider. There's already some code for this in SendDialog.setPaymentRequest.

If it's a pay-to-many transaction generated on the desktop, I think it's OK to just display a "Pay to many" placeholder in the "Pay to" box, as long as the total amount is correct.

Reply:
I have a question regarding this point. How would we extract the recipient address and amount if loading a fully signed unrelated transaction?

For related transactions, it would be doable by reading the `amount` field from `txInfo`, or summing up non-change address outputs, I gathered. However, can we get the correct receiver and amount from an unrelated transaction? I tried this on desktop (signed a transaction on a standard wallet and tried to broadcast it using a multisig wallet) and I get this:

 ![image](https://user-images.githubusercontent.com/24822460/121403024-cab48580-c95a-11eb-9c95-6f9ac1801dbf.png)

Seems there is no indicator which of the addresses is the receiving one. What can we do about situations like this?

Reply:
Yes, if the transaction is fully signed and unrelated then there's no need to show the details. In fact, I don't even think we need to show the details for a fully signed *related* transaction, though you can still do that if you want. The only case where the user must see the details is where they're adding a signature.

Reply:
I understand, and I implemented it as such in the latest commits.

I still have two questions regarding this point that we can maybe discuss.

1. This might be something that is never going to happen in a real-life scenario, but should we treat sending to addresses belonging the same wallet as a "distinct" output? That is, include it in the total amount and similar. As far as I could notice, the desktop version does not do that in the `Amount` field, only displaying the amount that is actually leaving the wallet.

2. I went with your suggestion, and implemented the details dialog when signing a transaction, not when sending it. However, I'm wondering if I should make it so that the last signature _immediately_ broadcasts the transaction, too. Currently, if yours is the last signature, the signed string will be copied into `ColdLoad` dialog, and the user can simply click `Send` to broadcast it. But I'm wondering if I should streamline that process. One reason why I didn't do it in the first place was because I figured that some users might not want to immediately broadcast, but instead keep the fully signed transaction "cold" for some time. What is your opinion on this?

Reply:
> should we treat sending to addresses belonging the same wallet as a "distinct" output? That is, include it in the total amount and similar. As far as I could notice, the desktop version does not do that in the Amount field, only displaying the amount that is actually leaving the wallet.  Currently, I'm using the wallet.is_mine method to disregard change addresses (so that it could also work in wallets that don't have change addresses enabled), but should we care about this?

I think it's fine the way it is, as the amount plus fee is consistent with the value which will eventually appear in the transaction list.

> I'm wondering if I should make it so that the last signature immediately broadcasts the transaction, too. [...] One reason why I didn't do it in the first place was because I figured that some users might not want to immediately broadcast, but instead keep the fully signed transaction "cold" for some time.

Yes, I think we should support that, so the way you've implemented it is fine.